Benchmark/src/PWR022.cpp

```diff
@@ -0,0 +1,72 @@
+#include "Benchmark.h"
+
+// Forward-declare the functions to benchmark
+extern "C" {
+void example(int n, const double *A, const double *B, const double *C,
+             double *D);
+void solution(int n, const double *A, const double *B, const double *C,
+              double *D);
+void example_f(int n, double *A, double *B, double *C, double *D);
+void solution_f(int n, double *A, double *B, double *C, double *D);
+}
+
+// Size adjusted to fit execution on micro-seconds
+constexpr int N = 1024;
```

Comment:
Isn't the size too large? Doing some rough measurements, the `N = 1024` configuration results in pretty large vectors. For instance, `C` reaches 8 GB!

```
C = 1024 * 1024 * 1024 * 8 bytes/double
```

I've run the benchmark and, based on what I found on my system, the amount of iterations for this benchmark is quite low, compared to other checks. For instance:

```
------------------------------------------------------------------
Benchmark                        Time             CPU   Iterations
------------------------------------------------------------------
PWR039 C Example              1062 us         1061 us          662
PWR039 C Improved              351 us          351 us         1978
PWR039 Fortran Example         919 us          919 us          762
PWR039 Fortran Improved        261 us          261 us         2693
```

Whereas PWR022 records just 2 iterations:

```
gcc/g++/gfortran (Debian 12.2.0-14) 12.2.0
------------------------------------------------------------------
Benchmark                        Time             CPU   Iterations
------------------------------------------------------------------
PWR022 C Example            447906 us       447685 us            2
PWR022 C Improved           458932 us       458162 us            2
PWR022 Fortran Example      451947 us       451172 us            2
PWR022 Fortran Improved     468787 us       468772 us            2
```

This makes sense, given the large amount of data processed. As can be seen in the table above, I'm also having a hard time obtaining any speedup from the benchmark with `gcc`. I also cannot claim that `clang` is in a better position:

```
Debian clang version 14.0.6
Target: x86_64-pc-linux-gnu
------------------------------------------------------------------
Benchmark                        Time             CPU   Iterations
------------------------------------------------------------------
PWR022 C Example            441338 us       440679 us            2
PWR022 C Improved           438093 us       438082 us            2
```

By reducing the problem size, I obtain much better results. The increased number of iterations must definitely help to obtain much more stable and reliable measures. For instance, using `N = 128` (`C` reaches just 16 MB):

```
------------------------------------------------------------------
Benchmark                        Time             CPU   Iterations
------------------------------------------------------------------
PWR022 C Example               475 us          475 us         1474
PWR022 C Improved              478 us          478 us         1482
PWR022 Fortran Example         500 us          499 us         1407
PWR022 Fortran Improved        439 us          439 us         1588
```

Now the Fortran example is measuring a noticeable speedup. Still, it is not as high as in the benchmarks of the other examples. Does the Fortran compiler manage to include vector instructions due to the PWR022 refactorization? Moreover, the C example shows no benefit. I think we need to dig deeper into these issues.

What results are you obtaining in your system?

Reply:
Thank you for the detailed response :)

The results from the Intel compilers look good indeed!

Out of curiosity, I have compared the assembly produced by `gcc 12.2`, `clang 14.0`, and `icc 2021.2` for the optimized C example. I haven't reviewed the code in detail because it spans hundreds of instructions, but I've been able to see that `icc` has a much more heavy usage of vector instructions compared to the other compilers. This suggests that the Intel compilers are indeed able to perform a better flow analysis as you pointed out, and therefore produce a more optimized assembly.

Even though `icc` and `ifort` achieve good speedups in this benchmark, I think that it would be convenient to update the example in the future to enable both `gcc` and `clang` to show decent speedups. At the end of the day, it is more likely that people use those compilers instead of Intel's.

Let's proceed on merging this PR, and leave this thread open as a reminder. WDYT?

Reply:
After giving it a second thought, you're definitely right. We shouldn't strive to make examples tailored to particular compilers. On the contrary, this type of finding is interesting to compare the capabilities of different compilers.

> I believe we should assert the fact that different compilers may show better or worse results in the benchmarks depending on their underlying implementation, yet that does not render the check obsolete.

+1

When I proposed this change, I was thinking about the possibility that another user in the future runs the benchmark and also does not obtain speedup, and starts wondering to what extent the check has a real impact. Even if a coding best practice is right in theory, I believe it is still important to find at least one situation (i.e., example code + compiler) where the check allows for a performance gain; in this case, the results of `icc` are perfect. This would definitely motivate the developer to adopt the best practice, as it verifies that the check has a current performance impact, and potential to be even more valuable in the future if more compilers learn to take better advantage of it.

In this line, the solution would rather be to evaluate the checks from the catalog against a suite of compilers, and report the results for quick comparison. Naturally, it is important that every user evaluates the benchmarks on their development platform for a more accurate assessment, but there would be great value in providing the community with this type of information to avoid local management of the installation of many compilers to compare them. Anyways, this is a topic for further review in the future, outside of the scope of this check.

Thanks for this interesting discussion :)